src/compiler/core.ts

```diff
@@ -1966,7 +1966,9 @@ namespace ts {
         const moduleKind = getEmitModuleKind(compilerOptions);
         return compilerOptions.allowSyntheticDefaultImports !== undefined
             ? compilerOptions.allowSyntheticDefaultImports
-            : moduleKind === ModuleKind.System;
+            : compilerOptions.ESModuleInterop
+                ? !!moduleKind && moduleKind < ModuleKind.ES2015

```

Comment:
nit `moduleKind !== ModuleKind.None && moduleKind < ModuleKind.ES2015` is easier to read.

Reply:
Just to be pedantic: That doesn't handle if `compilerOptions.module` is `NaN`, but this does. 😛